### ISSUE TITLE
smb2_read_data fails on MacOS (10.15.7) shared files

### ISSUE THREAD

**Issue by toto263** [sahlberg/libsmb2#185]
smb2_read_data fails on macOS shares.
After some testing, I found that the error occurs at the followings.

I don't know why, but signature verification fails on mac, and it return "Wrong signature in received" error.
Remove the verification (memcmp), then works ok.

```
socket.c - smb2_read_data
    ...     
        /* We don't yet have the signing key until later, once session
         * setup has completed, so we can not yet verify the signature
         * of the final leg of session setup.
         */
        if (smb2->sign &&
            (smb2->hdr.flags & SMB2_FLAGS_SIGNED) &&
            (smb2->hdr.command != SMB2_SESSION_SETUP) ) {
                uint8_t signature[16];
                memcpy(&signature[0], &smb2->in.iov[1].buf[48], 16);
                if (smb2_calc_signature(smb2, &smb2->in.iov[1].buf[48],
                                        &smb2->in.iov[1],
                                        smb2->in.niov - 1) < 0) {
                        return -1;
                }

//--> works if commented out the followings
                if (memcmp(&signature[0], &smb2->in.iov[1].buf[48], 16)) {
                        smb2_set_error(smb2, "Wrong signature in received "
                                       "PDU");
                        return -1;
                }
        }

...
```

**Comment by sahlberg** [comment]
So when signing is on, the server accepts the signatures that libsmb2 sends to it  but libsmb2 rejects the signatures that mac sends to us?

Can you collect a network trace of when this fails as well as a network trace of when it works when you comment out the verification?
Can you also try forcing an older dialect in libsmb2 (you can set version via url argument)
for example try one of the 2.x dialects as well as one of the 3.x dialects (but not 3.1.1 which is where we have negotiation of signing algorithms)

**Comment by toto263** [comment]
I tried all dialects using "smb2_set_version(context, SMB2_VERSION_xxxx)", from SMB2_VERSION_0202 to SMB2_VERSION_0302, but failed.

I found that another signature verification in session_setup_cb() of libsmb2.c, and this works ok.
```
libsmb2.c - session_setup_cb()
    ...
                if (smb2->hdr.flags & SMB2_FLAGS_SIGNED) {
                        uint8_t signature[16];

                        memcpy(&signature[0], &smb2->in.iov[1].buf[48], 16);
                        if (smb2_calc_signature(smb2, &smb2->in.iov[1].buf[48],
                                                &smb2->in.iov[1],
                                                smb2->in.niov - 1) < 0) {
                                c_data->cb(smb2, -EINVAL, NULL, c_data->cb_data);
                                free_c_data(smb2, c_data);
                                return;
                        }
---> this verification is OK.
                        if (memcmp(&signature[0], &smb2->in.iov[1].buf[48], 16)) {
                                smb2_set_error(smb2, "Wrong signature in received "
                                               "PDU");
                                c_data->cb(smb2, -EINVAL, NULL, c_data->cb_data);
                                free_c_data(smb2, c_data);
                                return;
                        }
                }
```

Only smb2_read_data() function fails.

I am using this library on iOS, so I don't know how to collect network trace.

**Comment by toto263** [comment]
I think this is a same issue with issue #176.

**Comment by delins** [comment]
@toto263 A similar issue was fixed in #183. Could you check if the latest commit fixes this issue as well?

**Comment by toto263** [comment]
> @toto263 A similar issue was fixed in #183. Could you check if the latest commit fixes this issue as well?

I confirm that latest commit fixes this issue. Now working very well with macOS.
